### PR TITLE
Format function type (and others)

### LINKED ISSUE #4438: Formatting issue in 1.6
can be seen in 9a6db930b48b2e51216739aa0d9430e13121cc93

before (unformatted)
```typescript
function renderElement(
    element: Element,
    writer: upw.Writer,
    options: uh2p.Options,
    renderNode: (
        node: Node,
        parentBox: cb.Box,
        parentStyles: CSSStyleDeclaration,
        writer: upw.Writer,
        options: uh2p.Options
    ) => void
): void {
}
```

after formatting actual
```typescript
function renderElement(
    element: Element,
    writer: upw.Writer,
    options: uh2p.Options,
    renderNode: (
    node: Node,
    parentBox: cb.Box,
    parentStyles: CSSStyleDeclaration,
    writer: upw.Writer,
    options: uh2p.Options
    ) => void
): void {
}
```

after formatting expected
```typescript
function renderElement(
    element: Element,
    writer: upw.Writer,
    options: uh2p.Options,
    renderNode: (
        node: Node,
        parentBox: cb.Box,
        parentStyles: CSSStyleDeclaration,
        writer: upw.Writer,
        options: uh2p.Options
    ) => void
): void {
}
```

### PR DESCRIPTION
**Pull request by SaschaNaz**
Fixes #4438 ~~and (maybe) #4486~~.

Async-await current:
```typescript
async   function asyncFunction() {
    await
    Promise.resolve("await");
    return await   Promise.resolve("completed");
}
```

Fix:
```typescript
async function asyncFunction() {
    await
        Promise.resolve("await");
    return await Promise.resolve("completed");
}
```

Function type / Constructor type current:
```typescript
function renderElement(
    renderNode:
    (
    node: Node
    ) => void,
    newNode:
    new (
    name: string
    ) => Node
): void {
}
```

Fix:
```typescript
function renderElement(
    renderNode:
        (
            node: Node
        ) => void,
    newNode:
        new (
            name: string
        ) => Node
): void {
}
```

Parenthesized type current:
```typescript
function foo(
    bar:
    (
    boolean
    )
) {
}
```

Fix:
```typescript
function foo(
    bar:
        (
            boolean
        )
) {
}
```

Call/Constructor/Method signature current:
(Constructor signature only is covered)
```typescript
type Foo = {
    (
        call: any
    ): void;
    new (
    constr: any
    ): void;
    method(
        whatever: any
    ): void;
};
```

Fix:
```typescript
type Foo = {
    (
        call: any
    ): void;
    new (
        constr: any
    ): void;
    method(
        whatever: any
    ): void;
};
```

Tagged template string current:
```typescript
String.raw`foo`
String.raw  `bar${3}`
String.raw
`template`
```

Fix:
```typescript
String.raw `foo`
String.raw `bar${3}`
String.raw
    `template`
```

Type alias current:
```typescript
type   Alias =
number;
```

Fix:
```typescript
type Alias =
    number;
```

Type union current:
```typescript
type   Union = number  |  {}
|string
```

Fix:
```typescript
type Union = number | {}
    | string
```

- [x] Fix parenthesized type formatting
- [x] Add tests

### CODE REVIEW

File: src/services/formatting/smartIndenter.ts [hunk 1]
```diff
@@ -429,6 +429,7 @@ namespace ts.formatting {
                 case SyntaxKind.ArrayBindingPattern:
                 case SyntaxKind.ObjectBindingPattern:
                 case SyntaxKind.JsxElement:
+                case SyntaxKind.FunctionType:
```
Comment (mhegazy) [inline src/services/formatting/smartIndenter.ts]: we should also add:
- ConstructorType
- ConstructSignature
- CallSignature
- ParenthesizedType
- TaggedTemplateExpression
- AwaitExpression
- YieldExpression

Comment (mhegazy) [inline src/services/formatting/smartIndenter.ts]: and add tests for theses :)

**Comment by SaschaNaz** [comment]
@mhegazy I think most of the types are now covered except YieldExpression. Do you have an example that is not being properly formatted?

**Comment by mhegazy** [comment]
:+1:

**Comment by DanielRosenwasser** [comment]
Thanks @SaschaNaz!

@mhegazy should this be brought into 1.6 as well?

**Pull request merged**